components/brave_federated/communication_adapter.cc

```diff
@@ -0,0 +1,189 @@
+    const std::unique_ptr<std::string> response_body) {
+  bool failed_request =
+      !url_loader_->ResponseInfo() || !url_loader_->ResponseInfo()->headers;
+  reconnect_backoff_entry_->InformOfRequest(!failed_request);
+  int64_t reconnect_delay_in_seconds =
+      reconnect_backoff_entry_->GetTimeUntilRelease().InSeconds();
+
+  if (failed_request) {
+    VLOG(1) << "Failed to request tasks, retrying in "
+            << reconnect_delay_in_seconds << "s";
+    std::move(callback).Run({}, reconnect_delay_in_seconds);
+    return;
+  }
+
+  auto headers = url_loader_->ResponseInfo()->headers;
+  int response_code = headers->response_code();
+  // TODO(lminto): Check headers (as per Daniel instructions)
+  if (response_code == net::HTTP_OK) {
+    TaskList task_list = ParseTaskListFromResponseBody(*response_body);
+    bool empty_task_list = task_list.empty();
+    VLOG(2) << "Received " << task_list.size() << " tasks from FL service";
+
+    request_task_backoff_entry_->InformOfRequest(!empty_task_list);
+    int64_t request_task_delay_in_seconds =
+        request_task_backoff_entry_->GetTimeUntilRelease().InSeconds();
+    if (empty_task_list) {
+      VLOG(1) << "No tasks received from FL service, retrying in "
+              << request_task_delay_in_seconds << "s";
+      std::move(callback).Run({}, request_task_delay_in_seconds);
+      return;
+    }
+
+    std::move(callback).Run(task_list, request_task_delay_in_seconds);
+    return;
+  }
+
+  VLOG(1) << "Failed to request tasks. Response code: " << response_code;
+}
+
+void CommunicationAdapter::PostTaskResult(TaskResult result,
+                                          PostResultCallback callback) {
+  auto request = std::make_unique<network::ResourceRequest>();
+  request->url = GURL(features::GetFederatedLearningResultsEndpoint());
+  request->headers.SetHeader("Content-Type", "application/protobuf");
+  request->headers.SetHeader("Accept", "application/protobuf");
+  request->headers.SetHeader("X-Brave-FL-Federated-Learning", "?1");
+  request->credentials_mode = network::mojom::CredentialsMode::kOmit;
+  request->method = net::HttpRequestHeaders::kPostMethod;
+
+  VLOG(2) << "Posting Task results " << request->method << " " << request->url;
+
+  const std::string& payload = BuildPostTaskResultsPayload(result);
+
+  url_loader_ = network::SimpleURLLoader::Create(
+      std::move(request), GetNetworkTrafficAnnotationTag());
+  url_loader_->AttachStringForUpload(payload, "application/protobuf");
+  url_loader_->DownloadToStringOfUnboundedSizeUntilCrashAndDie(
+      url_loader_factory_.get(),
+      base::BindOnce(&CommunicationAdapter::OnPostTaskResult,
+                     base::Unretained(this), std::move(callback)));
+}
+
+void CommunicationAdapter::OnPostTaskResult(
+    PostResultCallback callback,
+    std::unique_ptr<std::string> response_body) {
+  if (!url_loader_->ResponseInfo() || !url_loader_->ResponseInfo()->headers) {
+    VLOG(1) << "Failed to post task results";
```

Comment:
It is okay that we don't fire the callback in this case?

Reply:
better to fire it